File: 98-validate-binary-search-tree/98-validate-binary-search-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
    return checkValid(root,-Infinity,Infinity)

};

function checkValid(root: TreeNode | null, left: number, right: number){
    if(root == null) return true
    
    if(root.val <= left || root.val >= right) return false
    
    return checkValid(root.left,left,root.val) && 
           checkValid(root.right,root.val,right)

}

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// recu(node,leftLimit,rightLimit)

class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidTree(root,Long.MIN_VALUE,Long.MAX_VALUE);
    }
    
    public boolean isValidTree(TreeNode root, long lLimit, long uLimit){
        
        if(root==null) return true;
        
        if(root.val<=lLimit || root.val>=uLimit) return false;
        
        
        //Lower Limit required for Right Side. 
        //Upper Limit required for Left Side
        
        boolean leftValid = isValidTree(root.left,lLimit,root.val);
        boolean rightValid = isValidTree(root.right,root.val,uLimit);
        
        return leftValid && rightValid;
    
            
            
    }
}

 



     